lib/srv/app/aws/assume_role.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/srv/app/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func updateAssumeRoleDuration(identity *tlsca.Identity, w http.ResponseWriter, req *http.Request, clock clockwork.Clock) error {
+	// Skip non-AssumeRole request
+	query, err := getAssumeRoleQuery(req)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	requestedDuration := getAssumeRoleDuration(query)
+	identityDuration := identity.Expires.Sub(clock.Now()).Round(time.Second)
+	switch {
+	// Deny access if identity duration is shorter than the minimum that can be
+	// requested.
+	case identityDuration < assumeRoleMinDuration:
+		return trace.AccessDenied("minimum AWS session duration is %v but Teleport identity expires in %v", assumeRoleMinDuration, identityDuration)
```

Comment:
This case is problematic. It will mean that 15 min before user cert expiration teleport  will not support STS assume role call. 

Does this error will be prorated the the user ? 
Can we just use the min time  in this case and ignore the STS assume role cycles ?

Reply:
The client side will get a 403 so at least it knows it is an access denied error but without looking at tsh debug log, the client wouldn't understand why it gets denied.

We are currently doing this for the AWS web console, see [discussion](https://github.com/gravitational/teleport/pull/9792#discussion_r787370342)

I think we should keep it consistent so either use the minimum or deny, but in both cases. What do you think?